### ISSUE TITLE
Weights at loss function

### ISSUE THREAD

**Issue by Ntiboroo1** [Shmuma/rl#1]
In train.py, if config.weight_samples==True #in 74. row
Instead of 'simple' value_loss_t = value_loss_t.mean(), shouldn't we divide value_loss_t.sum() by the sum of weights (i.e. weighted average).
I think we don't want the loss function to be lower just because the weights in the batch was lower, i.e. sampling 'farther' states.

**Comment by Shmuma** [comment]
Hm, from my perspective it looks ok. In the paper (https://arxiv.org/pdf/1805.07470.pdf, page 5, section "Weighted samples") they say that loss for every sample is weighted by 1/Depth. 

Tensor weight_t has the same dimension as our batch, so, by multiplying the loss we just weight every sample's loss value by their corresponding weight, decreasing the gradients for states which are further away from the solved state. 

Then, to make the batch loss independent from batch size, I take the mean, but it has no relation to weighted average, just a standard way of calculating the final loss from the batch. I guess the mean doesn't even affect the gradients, as the constant factor of the loss will be ignored by backpropagation.

**Comment by Ntiboroo1** [comment]
Thanks for the clarification and the quick answer,

You are right that the weighting only matters at decreasing the gradients for the further cube-states and have nothing to do with the weighted average [my bad, sorry for that].
Anyway I think the constant factor is important, however it will be multiplied by the learning rate, I think that was what you meant by ignored (?).

I have a new question regarding the calculation(?) of the loss,
As for the backpropagation part, I'm relatively new to DL:
Why do you need to calculate the loss **by hand** [or is it not really a calculation, but a formula for the pytorch], for backpropagation?
It will be the input to the .backward() method, which I don't really understand.
https://pytorch.org/docs/stable/autograd.html


In my understanding:
1. The type of the loss function used (L^2 error) is needed, and the labels for the points, so that the system knows how to calculate the gradient.
2. From that the gradient vector for the weights/biases for a single element in the minibatch is calculated
3. The mean of the gradients for the single elements is taken.
4. The system backpropagates the mean gradient vector from step 3, weights/biases are updated.

What do I see wrong? Can you please clarify this part for me?

**Comment by Shmuma** [comment]
> Anyway I think the constant factor is important, however it will be multiplied by the learning rate, I think that was what you meant by ignored (?).

Nope, if I remember correctly from cs231n course, if we're applying constant multiplier on whole batch, it doesn't change the scale of gradients. That's related to the way backpropagation works (taking the derivative of the function). On the other hand, learning rate does important, because it is being applied not on the backpropagation, but later, when we're updating the weights by adding scaled gradients (in the simplest SGD case).

> Why do you need to calculate the loss by hand [or is it not really a calculation, but a formula for the pytorch], for backpropagation?

From PyTorch point of view, both expressions below are equivalent:
```
value_loss_t = (value_out_t - y_value_t)**2
value_loss_t = torch.nn.MSELoss(reduction='none')(value_out_t, y_value_y)
```

And the latter, in fact, just slightly more complicated form of explicit MSE expression. That's true due to the way PyTorch tracks expressions and calculate gradients. For better understanding, I recommend check the tutorial from Jeremy Howard: https://pytorch.org/tutorials/beginner/nn_tutorial.html

Without the need of scaling samples' losses using weights, using `torch.nn.MSELoss` would be simpler, as by default, it calculates mean() automatically. But in this particular case, we need to scale, so, explicit formula is simpler, from my point of view.
